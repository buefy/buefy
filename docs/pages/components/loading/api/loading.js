export default [
    {
        props: [
            {
                name: '<code>active</code>',
                description: 'Whether modal is active or not, use the <code>.sync</code> modifier to make it two-way binding',
                type: 'Boolean',
                values: '—',
                default: '<code>false</code>'
            },
            {
                name: '<code>animation</code>',
                description: 'Custom animation (transition name)',
                type: 'String',
                values: '—',
                default: '<code>fade</code>'
            },
            {
                name: '<code>is-full-page</code>',
                description: 'Loader will overlay the full page',
                type: 'Boolean',
                values: '—',
                default: '<code>true</code>'
            },
            {
                name: '<code>canCancel</code>',
                description: `Can close Loading by pressing escape or clicking outside`,
                type: 'Boolean',
                values: '—',
                default: '<code>false</code>'
            },
            {
                name: '<code>onCancel</code>',
                description: `Callback function to call after user canceled (pressed escape / clicked outside)`,
                type: 'Function',
                values: '—',
                default: '—'
            }
        ],
        events: [
            {
                name: '<code>close</code>',
                description: 'Triggers when user closed/canceled or called programmatically from the injected component',
                parameters: '—'
            }
        ]
    }
]
