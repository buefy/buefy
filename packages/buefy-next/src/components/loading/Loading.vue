<template>
    <transition :name="animation">
        <div
            class="loading-overlay is-active"
            :class="{ 'is-full-page': displayInFullPage }"
            v-show="isActive"
            v-if="isActive"
        >
            <div class="loading-background" @click="cancel" />
            <slot>
                <div class="loading-icon" />
            </slot>
        </div>
    </transition>
</template>

<script>
import { removeElement } from '../../utils/helpers'
import { HTMLElement } from '../../utils/ssr'

export default {
    name: 'BLoading',
    props: {
        modelValue: Boolean,
        programmatic: Boolean,
        container: [Object, Function, HTMLElement],
        isFullPage: {
            type: Boolean,
            default: true
        },
        animation: {
            type: String,
            default: 'fade'
        },
        canCancel: {
            type: Boolean,
            default: false
        },
        onCancel: {
            type: Function,
            default: () => {}
        }
    },
    emits: ['close', 'update:is-full-page', 'update:modelValue'],
    data() {
        return {
            isActive: this.modelValue || false,
            displayInFullPage: this.isFullPage
        }
    },
    watch: {
        modelValue(value) {
            this.isActive = value
        },
        isFullPage(value) {
            this.displayInFullPage = value
        }
    },
    methods: {
        /**
        * Close the Modal if canCancel.
        */
        cancel() {
            if (!this.canCancel || !this.isActive) return

            this.close()
        },
        /**
        * Emit events, and destroy modal if it's programmatic.
        */
        close() {
            this.onCancel.apply(null, arguments)
            this.$emit('close')
            this.$emit('update:modelValue', false)

            // Timeout for the animation complete before destroying
            if (this.programmatic) {
                this.isActive = false
                // TODO: should the following happen outside this component;
                // i.e., in index.js?
                setTimeout(() => {
                    removeElement(this.$el)
                }, 150)
            }
        },
        /**
        * Keypress event that is bound to the document.
        */
        keyPress({ key }) {
            if (key === 'Escape' || key === 'Esc') this.cancel()
        }
    },
    created() {
        if (typeof window !== 'undefined') {
            document.addEventListener('keyup', this.keyPress)
        }
    },
    mounted() {
        // Insert the Loading component in body tag
        // only if it's programmatic
        // (moved from beforeMount because $el is not bound during beforeMount)
        // TODO: should this happen outside this component; i.e., in index.js?
        if (this.programmatic) {
            if (!this.container) {
                document.body.appendChild(this.$el)
            } else {
                this.displayInFullPage = false
                this.$emit('update:is-full-page', false)
                this.container.appendChild(this.$el)
            }
            this.isActive = true
        }
    },
    beforeUnmount() {
        if (typeof window !== 'undefined') {
            document.removeEventListener('keyup', this.keyPress)
        }
    }
}
</script>
