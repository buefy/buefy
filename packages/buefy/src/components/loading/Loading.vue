<template>
    <transition :name="animation">
        <div
            class="loading-overlay is-active"
            :class="{ 'is-full-page': displayInFullPage }"
            v-show="isActive"
            v-if="isActive"
        >
            <div class="loading-background" @click="cancel" />
            <slot>
                <div class="loading-icon" />
            </slot>
        </div>
    </transition>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import type { PropType } from 'vue'
import { removeElement } from '../../utils/helpers'
import type { ExtractComponentProps } from '../../utils/helpers'
import { HTMLElement } from '../../utils/ssr'

const Loading = defineComponent({
    name: 'BLoading',
    props: {
        modelValue: Boolean,
        programmatic: Boolean,
        container: [Object, Function, HTMLElement] as PropType<HTMLElement>,
        isFullPage: {
            type: Boolean,
            default: true
        },
        animation: {
            type: String,
            default: 'fade'
        },
        canCancel: {
            type: Boolean,
            default: false
        },
        onCancel: {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            type: Function as PropType<(...args: any[]) => void>,
            default: () => {}
        }
    },
    emits: {
        /* eslint-disable @typescript-eslint/no-unused-vars */
        close: () => true,
        'update:is-full-page': (_flag: boolean) => true,
        'update:modelValue': (_flag: boolean) => true
        /* eslint-enable @typescript-eslint/no-unused-vars */
    },
    data() {
        return {
            isActive: this.modelValue || false,
            displayInFullPage: this.isFullPage
        }
    },
    watch: {
        modelValue(value) {
            this.isActive = value
        },
        isFullPage(value) {
            this.displayInFullPage = value
        }
    },
    methods: {
        /*
        * Close the Modal if canCancel.
        */
        cancel() {
            if (!this.canCancel || !this.isActive) return

            this.close()
        },
        /*
        * Emit events, and destroy modal if it's programmatic.
        */
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        close(...args: any[]) {
            this.onCancel.apply(null, args)
            this.$emit('close')
            this.$emit('update:modelValue', false)

            // Timeout for the animation complete before destroying
            if (this.programmatic) {
                this.isActive = false
                // TODO: should the following happen outside this component;
                // i.e., in index.ts?
                setTimeout(() => {
                    removeElement(this.$el)
                }, 150)
            }
        },
        /*
        * Keypress event that is bound to the document.
        */
        keyPress({ key }: KeyboardEvent) {
            if (key === 'Escape' || key === 'Esc') this.cancel()
        }
    },
    created() {
        if (typeof window !== 'undefined') {
            document.addEventListener('keyup', this.keyPress)
        }
    },
    mounted() {
        // Insert the Loading component in body tag
        // only if it's programmatic
        // (moved from beforeMount because $el is not bound during beforeMount)
        // TODO: should this happen outside this component; i.e., in index.js?
        if (this.programmatic) {
            if (!this.container) {
                document.body.appendChild(this.$el)
            } else {
                this.displayInFullPage = false
                this.$emit('update:is-full-page', false)
                this.container.appendChild(this.$el)
            }
            this.isActive = true
        }
    },
    beforeUnmount() {
        if (typeof window !== 'undefined') {
            document.removeEventListener('keyup', this.keyPress)
        }
    }
})

export type LoadingProps = ExtractComponentProps<typeof Loading>

export default Loading
</script>
