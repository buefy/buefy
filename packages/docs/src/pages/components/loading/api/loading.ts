export default [
    {
        props: [
            {
                name: '<code>v-model</code>',
                description: 'Whether loading is active or not',
                type: 'Boolean',
                values: '—',
                default: '<code>false</code>'
            },
            {
                name: '<code>programmatic</code>',
                description: 'Indicates if the loading was created programmatically',
                type: 'Boolean',
                values: '—',
                default: '<code>false</code>'
            },
            {
                name: '<code>container</code>',
                description: 'DOM element where the loading component will be created on (for programmatic usage)',
                type: 'Object, Function, HTMLElement',
                values: '—',
                default: '<code>document.body</code>'
            },
            {
                name: '<code>animation</code>',
                description: 'Custom animation (transition name)',
                type: 'String',
                values: '—',
                default: '<code>fade</code>'
            },
            {
                name: '<code>is-full-page</code>',
                description: 'Loader will overlay the full page',
                type: 'Boolean',
                values: '—',
                default: '<code>true</code>'
            },
            {
                name: '<code>can-cancel</code>',
                description: `Can close Loading by pressing escape or clicking outside`,
                type: 'Boolean',
                values: '—',
                default: '<code>false</code>'
            },
            {
                name: '<code>on-cancel</code>',
                description: `Callback function to call after user canceled (pressed escape / clicked outside)`,
                type: 'Function',
                values: '—',
                default: '—'
            }
        ],
        slots: [
            {
                name: 'default',
                description: 'Loading icon',
                props: '—'
            },
        ],
        events: [
            {
                name: '<code>close</code>',
                description: 'Triggers when user closed/canceled or called programmatically from the injected component',
                parameters: '—'
            },
            {
                name: '<code>update:modelValue</code>',
                description: 'Triggers when the loading state changes',
                parameters: '<code>value: Boolean</code>'
            },
            {
                name: '<code>update:is-full-page</code>',
                description: 'Triggers when the full-page state changes',
                parameters: '<code>value: Boolean</code>'
            },
        ],
        methods: [
            {
                name: '<code>close</code>',
                description: 'Programmatically close the loading overlay',
                return: '—'
            }
        ]
    }
]
