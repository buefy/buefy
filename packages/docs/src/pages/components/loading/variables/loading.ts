export default [
    {
        sass: '<code>$loading-background</code>',
        css: '<code>--bulma-loading-background</code>',
        description: 'The loading background color',
        default: '<code>rgba($scheme-main, 50%)</code>'
    },
    {
        sass: '<code>$loading-icon-size</code>',
        css: '<code>--bulma-loading-icon-size</code>',
        description: 'The loading icon size',
        default: '<code>3em</code>'
    },
    {
        sass: '<code>$loading-full-page-icon-size</code>',
        css: '<code>--bulma-loading-full-page-icon-size</code>',
        description: 'The loading icon size when loader is displayed over full page',
        default: '<code>5em</code>'
    }
]
