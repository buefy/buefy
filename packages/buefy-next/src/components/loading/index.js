import { createApp, h as createElement } from 'vue'

import Loading from './Loading.vue'

import { merge, copyAppContext, getComponentFromVNode } from '../../utils/helpers'
import { use, registerComponent, registerComponentProgrammatic } from '../../utils/plugins'

class LoadingProgrammatic {
    constructor(app) {
        this.app = app // may be undefined in the testing environment
    }

    open(params) {
        const defaultParam = {
            programmatic: true
        }
        const propsData = merge(defaultParam, params)
        const container = document.createElement('div')
        // Vue 3 requires a new app to mount another component
        const vueInstance = createApp({
            data() {
                return {
                    loadingVNode: null
                }
            },
            methods: {
                close() {
                    // TODO: too much dependence on Vue's internal structure?
                    const loading = getComponentFromVNode(this.loadingVNode)
                    if (loading) {
                        loading.close()
                    }
                }
            },
            render() {
                this.loadingVNode = createElement(
                    Loading,
                    {
                        ...propsData,
                        onClose(...args) {
                            if (propsData.onClose) {
                                propsData.onClose(...args)
                            }
                            // timeout for the animation complete before destroying
                            setTimeout(() => {
                                vueInstance.unmount()
                            }, 150)
                        }
                    }
                )
                return this.loadingVNode
            }
        })
        if (this.app) {
            copyAppContext(this.app, vueInstance)
        }
        return vueInstance.mount(container)
    }
}

const Plugin = {
    install(Vue) {
        registerComponent(Vue, Loading)
        registerComponentProgrammatic(Vue, 'loading', new LoadingProgrammatic(Vue))
    }
}

use(Plugin)

export default Plugin

export {
    LoadingProgrammatic,
    Loading as BLoading
}
