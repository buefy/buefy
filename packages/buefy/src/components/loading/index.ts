import { createApp, h as createElement } from 'vue'
import type { App, ComponentPublicInstance } from 'vue'

import Loading from './Loading.vue'
import type { LoadingProps } from './Loading.vue'

import { copyAppContext, getComponentFromVNode } from '../../utils/helpers'
import { registerComponent, registerComponentProgrammatic } from '../../utils/plugins'

export type LoadingOpenParams = Omit<LoadingProps, 'programmatic'>

// Minimal definition of a programmatically opened Loading.
//
// ESLint does not like `{}` as a type but allowed here to make them look
// similar to Vue's definition: https://github.com/vuejs/core/blob/fbc0c42bcf6dea5a6ae664223fa19d4375ca39f0/packages/runtime-core/src/componentPublicInstance.ts#L290-L305.
/* eslint-disable @typescript-eslint/ban-types */
type LoadingProgrammaticInstance = ComponentPublicInstance<
    {}, // P
    {}, // B
    {}, // D
    {}, // C
    { close: () => void } // M
>
/* eslint-enable @typescript-eslint/ban-types */

class LoadingProgrammatic {
    private app: App | undefined

    constructor(app?: App) {
        this.app = app // may be undefined in the testing environment
    }

    open(params: LoadingOpenParams) {
        const propsData = params
        const container = document.createElement('div')
        // Vue 3 requires a new app to mount another component
        const vueInstance = createApp({
            data() {
                return {
                    loadingVNode: null
                }
            },
            methods: {
                close() {
                    // TODO: too much dependence on Vue's internal structure?
                    const loading = getComponentFromVNode(this.loadingVNode)
                    if (loading) {
                        (loading as LoadingProgrammaticInstance).close()
                    }
                }
            },
            render() {
                this.loadingVNode = createElement(
                    Loading,
                    {
                        ...propsData,
                        programmatic: true,
                        onClose(...args) {
                            if (propsData.onClose) {
                                propsData.onClose(...args)
                            }
                            // timeout for the animation complete before destroying
                            setTimeout(() => {
                                vueInstance.unmount()
                            }, 150)
                        }
                    }
                )
                return this.loadingVNode
            }
        })
        if (this.app) {
            copyAppContext(this.app, vueInstance)
        }
        return vueInstance.mount(container) as LoadingProgrammaticInstance
    }
}

const Plugin = {
    install(Vue: App) {
        registerComponent(Vue, Loading)
        registerComponentProgrammatic(Vue, 'loading', new LoadingProgrammatic(Vue))
    }
}

export default Plugin

export {
    LoadingProgrammatic,
    Loading as BLoading
}
