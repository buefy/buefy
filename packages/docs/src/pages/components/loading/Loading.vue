<template>
    <div>
        <Example :component="ExSimple" :code="ExSimpleCode">
            <p>The Loading will be closed after about 10 seconds, by pressing escape or by clicking outside.</p>
            <p>
                Use <code>:is-full-page="false"</code> to limit loader to its container.<br>
                In this case, the container element should be positioned as <code>position: relative</code>.
            </p>
        </Example>

        <Example :component="ExProgrammatically" :code="ExProgrammaticallyCode" title="Programmatically opening">
            <p>When you want to close the Loading, call the <code>close()</code> method from the component.</p>
        </Example>

        <Example :component="ExTemplated" :code="ExTemplatedCode" title="Templated">
            <div class="tags has-addons">
                <span class="tag is-success">Since</span>
                <span class="tag is-info">0.7.5</span>
            </div>
            <p>Slot is available for loading content.</p>
        </Example>

        <ApiView :data="api"/>
        <VariablesView :data="variables"/>
    </div>
</template>

<script lang="ts">
    import { defineComponent } from 'vue'

    import { shallowFields } from '@/utils'
    import ApiView from '@/components/ApiView.vue'
    import Example from '@/components/Example.vue'
    import VariablesView from '@/components/VariablesView.vue'

    import api from './api/loading'
    import variables from './variables/loading'

    import ExSimple from './examples/ExSimple.vue'
    import ExSimpleCode from './examples/ExSimple.vue?raw'

    import ExProgrammatically from './examples/ExProgrammatically.vue'
    import ExProgrammaticallyCode from './examples/ExProgrammatically.vue?raw'

    import ExTemplated from './examples/ExTemplated.vue'
    import ExTemplatedCode from './examples/ExTemplated.vue?raw'

    export default defineComponent({
        components: {
            ApiView,
            Example,
            VariablesView
        },
        data() {
            return {
                api,
                variables,
                ...shallowFields({
                    ExSimple,
                    ExProgrammatically,
                    ExTemplated
                }),
                ExSimpleCode,
                ExProgrammaticallyCode,
                ExTemplatedCode
            }
        }
    })
</script>
