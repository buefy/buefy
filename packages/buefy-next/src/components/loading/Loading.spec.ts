import { transformVNodeArgs } from 'vue'
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'
import type { Mock, MockInstance } from 'vitest'
import { shallowMount } from '@vue/test-utils'
import type { VueWrapper } from '@vue/test-utils'
import { BLoading, LoadingProgrammatic } from '@components/loading'

type LoadingProgrammaticInstance = ReturnType<typeof LoadingProgrammatic.prototype.open>

let wrapper: VueWrapper<InstanceType<typeof BLoading>>

describe('BLoading', () => {
    beforeEach(() => {
        wrapper = shallowMount(BLoading)
    })

    it('is inactive initially', () => {
        expect(wrapper.vm.isActive).toBeFalsy()
    })

    describe('active', () => {
        beforeEach(() => {
            wrapper.setProps({ modelValue: true })
        })

        it('Is called', () => {
            expect(wrapper.vm).toBeTruthy()
            expect(wrapper.vm.$options.name).toBe('BLoading')
        })

        it('render correctly', () => {
            expect(wrapper.html()).toMatchSnapshot()
        })

        it('changes isActive when active prop is modified', async () => {
            await wrapper.setProps({ modelValue: false })
            expect(wrapper.vm.isActive).toBeFalsy()
            await wrapper.setProps({ modelValue: true })
            expect(wrapper.vm.isActive).toBeTruthy()
        })

        it('close on cancel', async () => {
            await wrapper.setProps({ canCancel: true })
            await wrapper.setData({ isActive: true })
            wrapper.vm.close = vi.fn()
            wrapper.vm.cancel()
            await wrapper.vm.$nextTick()
            expect(wrapper.vm.close).toHaveBeenCalled()
        })

        it('close on escape', async () => {
            await wrapper.setData({ isActive: true })
            wrapper.vm.cancel = vi.fn(() => wrapper.vm.cancel)
            const event = new KeyboardEvent('keyup', { key: 'Escape' })
            wrapper.vm.keyPress({} as KeyboardEvent)
            wrapper.vm.keyPress(event)
            expect(wrapper.vm.cancel).toHaveBeenCalledTimes(1)
        })

        it('emit events on close', () => {
            wrapper.vm.close()
            expect(wrapper.emitted().close).toBeTruthy()
            expect(wrapper.emitted()['update:modelValue']).toBeTruthy()
        })
    })

    describe('programmatic without container', () => {
        let loading: LoadingProgrammaticInstance
        let onClose: Mock
        let spyOnAppendChild: MockInstance

        beforeEach(async () => {
            // resets stubs introduced by @vue/test-utils
            // otherwise, every BLoading becomes a stub
            transformVNodeArgs()

            onClose = vi.fn()
            spyOnAppendChild = vi.spyOn(window.document.body, 'appendChild')
            loading = new LoadingProgrammatic().open({ onClose })
            await loading.$nextTick() // makes sure DOM is updated
        })

        afterEach(() => {
            spyOnAppendChild.mockRestore()
            vi.useFakeTimers()
            loading.close()
            // subsequent tests will fail unless we wait for the timeout
            vi.advanceTimersByTime(150)
            vi.useRealTimers()
        })

        it('Is called', () => {
            expect(spyOnAppendChild).toHaveBeenCalled()
            expect(window.document.querySelector('.loading-overlay')).toBeTruthy()
        })

        it('Is full page', () => {
            expect(window.document.querySelector('.is-full-page')).toBeTruthy()
        })

        it('manage close', async () => {
            vi.useFakeTimers()
            vi.spyOn(global, 'setTimeout')

            loading.close()
            expect(onClose).toHaveBeenCalled()
            expect(window.document.querySelector('.loading-overlay')).toBeTruthy()
            expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 150)

            vi.advanceTimersByTime(150)
            expect(window.document.querySelector('.loading-overlay')).toBeFalsy()
        })
    })

    describe('programmatic with a container', () => {
        let loading: LoadingProgrammaticInstance
        let container: HTMLElement
        let spyOnAppendChild: MockInstance

        beforeEach(async () => {
            // resets stubs introduced by @vue/test-utils
            // otherwise, every BLoading becomes a stub
            transformVNodeArgs()

            container = document.createElement('div')
            spyOnAppendChild = vi.spyOn(container, 'appendChild')
            loading = new LoadingProgrammatic().open({
                container
            })
            await loading.$nextTick() // makes sure DOM is updated
        })

        afterEach(() => {
            spyOnAppendChild.mockRestore()
            vi.useFakeTimers()
            loading.close()
            // subsequent tests may fail unless we wait for the timeout
            vi.advanceTimersByTime(150)
            vi.useRealTimers()
        })

        it('Is called', () => {
            expect(container.querySelector('.loading-overlay')).toBeTruthy()
            expect(spyOnAppendChild).toHaveBeenCalled()
        })

        it('Is not full page', () => {
            expect(container.querySelector('.is-full-page')).toBeFalsy()
        })
    })
})
