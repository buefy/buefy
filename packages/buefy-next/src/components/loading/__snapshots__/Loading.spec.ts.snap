// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BLoading > active > render correctly 1`] = `
"<transition-stub name=\\"fade\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
  <div class=\\"loading-overlay is-active is-full-page\\">
    <div class=\\"loading-background\\"></div>
    <div class=\\"loading-icon\\"></div>
  </div>
</transition-stub>"
`;
